invoicenet/acp/data.py

```diff
@@ -124,7 +124,7 @@ def _encode_ngrams(self, n_grams, height, width):
             bottom = min(round(word['bottom'] * v_ar), max_v)
 
             for k, v in n_gram['parses'].items():
-                parses[top:bottom, left:right, m_idx, self.parses_idx[k]] = 1.0
+                parses[top:bottom + 1, left:right + 1, m_idx, self.parses_idx[k]] = 1.0
```

Comment:
Am I right to think that right + 1 and bottom + 1 could exceed the range in some extreme cases for words that happen to end at the extreme right or bottom of the page? In that case right + 1 = max_h and bottom + 1 = max_v and we'd end up hitting an index out of range error. These changes LGTM but I'd suggest adding a check to cover the edge cases.

Reply:
Changing the limits on lines 121-124 should work I think

```python
left = min(round(word['left'] * h_ar), max_h - 1)
right = min(round(word['right'] * h_ar), max_h - 1)
top = min(round(word['top'] * v_ar), max_v - 1)
bottom = min(round(word['bottom'] * v_ar), max_v - 1)
```

Reply:
We checked the implementation and indexing already does this checks both when used with python lists and also numpy arrays.
`x = [[1,2,3], [4,5,6]]
print(x[0:10][0:10])`
Output: [[1,2,3], [4,5,6]]
`x = np.array([[1,2,3], [1,2,3]])
print(x[0:10,0:10].size)`
Output: [[1,2,3], 
 [4,5,6]]

Your suggested solution seems to avoid using right outer column and bottom outer row.

Reply:
np.linspace should also work just fine, because we are setting the boundaries to be 1 less then the size of the image.
`max_v = self.im_size[0] - 1`
`max_h = self.im_size[1] - 1`

Reply:
Ah, okay perfect!

Reply:
There is only a need to cover the errors possibly caused by exact indexing. (in the case of invoicenet - char_indices). 
`x = np.array([[1,2,3], [1,2,3]])`
`print(x[1:3, 3])`
IndexError: index 3 is out of bounds for axis 1 with size 3

But as I mentioned in the previous message, we are avoiding this edge case by setting the max_v and max_h values to im_size - 1.